src/Interpreters/ClusterProxy/executeQuery.cpp

```diff
@@ -189,7 +192,22 @@ void executeQuery(
             visitor.visit(query_ast_for_shard);
         }
         else
-            query_ast_for_shard = query_ast;
+            query_ast_for_shard = query_ast->clone();
+
+        if (shard_filter_generator)
+        {
+            auto shard_filter = shard_filter_generator(shard_info.shard_num);
+            if (shard_filter)
+            {
+                auto & select_query = query_ast_for_shard->as<ASTSelectQuery &>();
+
+                auto where_expression = select_query.where();
+                if (where_expression)
+                    shard_filter = makeASTFunction("and", where_expression, shard_filter);
```

Comment:
Now I think it is not the best approach (probably, will be had to implement with analyzer).
Note, that additional_tables_filter is implemented in a different way. It is propagated through settings (see updateSettingsForCluster), and the condition is added later in Interpreter.
https://github.com/ClickHouse/ClickHouse/blob/6a086ab9264f6a2ce1adefa6f204f76220a9d015/src/Interpreters/InterpreterSelectQuery.cpp#L1439-L1449

Probably we can do something similar.

Reply:
Probably we can add it if `parallel_replicas_count > 0`

Reply:
Not sure how it's connected to the new analyzer because this part is storage related right before we read from the remote replica.
The thing that doesn't work with the analyzer is setting `distributed_group_by_no_merge` to `2` and needs to be added there (will do it in a separate PR).
I can do it that way, I would need to set the settings `parallel_replicas_count` manually if `query_info.use_custom_key` is `true`. 

>  and the condition is added later in Interpreter

I already do this for cases where we have multiple shards so I don't need to add extra code just set `parallel_replicas_offset/count`

Reply:
All right.
My reasoning was:
* With analyzer, you may not have an AST at all. But probably I am wrong and precisely in this place we would have it - till we have to send query text for distributed processing
* Generally, appending something to `WHERE` is not the same as filtering just after reading. An example is `JOIN` : `SELECT * FROM table_1 RIGHT JOIN table_2 USING (x) WHERE x > 0` != `SELECT * FROM (SELECT * FROM table_1 WHERE x > 0) RIGHT JOIN table_2 USING (x)` . But, again, probably it's file as long as the feature is disabled for JOINs

Reply:
So, may be it's ok for now.

Reply:
I see, I disabled it for JOINs (and added test), and for now I don't see a way of using it with custom key.
Tried adding `parallel_replicas_offset` but turned out a bit more complex because I would need to copy context for each shard and figure out how to assign offsets.
But as I said, I have the logic for handling `parallel_replicas_offset/count` together with custom key (creates a filter in the same way as `additional_tables_filter`)  so I can easily switch to it in the future PRs just by setting those settings and remove appending `WHERE`.